cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/AbstractMDSUpgrader.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.datafabric.dataset.service.mds;
+
+import co.cask.cdap.common.io.Locations;
+import com.google.common.collect.ImmutableList;
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * An Abstract class which provides common methods and classes need by Dataset MDS to be upgraded
+ */
+public abstract class AbstractMDSUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractMDSUpgrader.class);
+
+
+  /**
+   * An abstract method which should be implemented by dataset MDS which needs to be upgraded
+   * @throws Exception
+   */
+  public abstract void upgrade() throws Exception;
+
+  /**
+   * Renames the old location to new location if old location exists and the new one does not
+   *
+   * @param oldLocation the old {@link Location}
+   * @param newLocation the new {@link Location}
+   * @return new location if and only if the file or directory is successfully moved; null otherwise.
+   * @throws IOException
+   */
+  @Nullable
+  protected Location renameLocation(Location oldLocation, Location newLocation) throws IOException {
+    // if the newLocation does not exists or the oldLocation does we try to rename. If either one of them is false then
+    // the underlying call to renameTo will throw IOException which we propagate.
+    if (!newLocation.exists() || oldLocation.exists()) {
```

Comment:
There is also the case where two file sets share the same base path. This could be due to the user who created them specifying the same path, or it could be due to the bug https://issues.cask.co/browse/CDAP-1721. What would happen? The first file set gets moved, and the old directory now does not exist any more. When the second fileset gets upgraded, we want to detect that, and somehow log it, but we don't want the upgrade tool to fail because of that.